xmtp_mls/src/groups/mod.rs

```diff
@@ -774,6 +778,39 @@ fn build_group_config(
         .build())
 }
 
+async fn validate_initial_group_membership<ApiClient: XmtpApi>(
+    client: &Client<ApiClient>,
+    conn: &DbConnection,
+    mls_group: &OpenMlsGroup,
+) -> Result<(), GroupError> {
+    let membership = extract_group_membership(mls_group.extensions())?;
+    let needs_update = client.filter_inbox_ids_needing_updates(conn, membership.to_filters())?;
+    if !needs_update.is_empty() {
+        load_identity_updates(&client.api_client, conn, needs_update).await?;
+    }
+
+    let mut expected_installation_ids = HashSet::<Vec<u8>>::new();
+
+    for (inbox_id, sequence_id) in membership.members {
+        let association_state = client
+            .get_association_state(conn, inbox_id, Some(sequence_id as i64))
+            .await?;
+
+        expected_installation_ids.extend(association_state.installation_ids());
+    }
+
+    let actual_installation_ids: HashSet<Vec<u8>> = mls_group
+        .members()
+        .map(|member| member.signature_key)
+        .collect();
+
+    if expected_installation_ids != actual_installation_ids {
```

Comment:
I've left this as a strict validation where the expected and actual memberships need to match exactly.

There is a looser form where we just validate that every actual installation is allowed to be there, which would allow for cases where key packages were broken or invalid at the time the user was added to the group. I think this is the right call: we should think about the "installations who should be in the group but can't" problem more holistically.